hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java

```diff
@@ -471,4 +444,37 @@ public ContainerStateManager getContainerStateManager() {
   public SCMHAManager getSCMHAManager() {
     return haManager;
   }
+
+  private static List<ContainerID> filterSortAndLimit(
```

Comment:
I think this could be made more efficient. The current algorithm takes the set, forms a new list of all elements. Then filters, sorted and returns the range.

Using an adaption of the priorityQueue technique here - https://www.baeldung.com/java-array-top-elements

We could avoid the copy from set to list, and the potentially large sort.

```
for (contaienrID cid : set) {
    if (id > startID)
        maxHeap.add(number);

        if (maxHeap.size() > count) {
            maxHeap.poll();
        }
    }
    return new ArrayList<>(maxHeap);
    // Might need reversed, I am not sure
});
```

I think this would be more memory efficient on large container sets, and should perform better than the full sort

Reply:
Do we even need to sort the list? The interface javadoc does not say that the returned list will be sorted. I know the earlier implementation is sorting, and there might be other code that's already relying on a sorted order... But I think it should be the caller's responsibility to manipulate the data however they want.

Reply:
I wonder how many people have not noticed that issue, and just implemented the linked solution as is!